L9-go-resolver-balancer-API.md

```diff
@@ -0,0 +1,504 @@
+Resolver and balancer API changes
+----
+* Author(s): menghanl, dfawley, mmukhi
+* Approver: a11r
+* Status: In Review
+* Implemented in: <language, ...>
+* Last updated: 2017/07/19
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/rUDN7-cZvz8
+
+## Abstract
+
+The current balancer API in gRPC-go cannot suport balancer hot switching. There are also other issues related to balancer, resolver and connectivity state.
+
+This document contains the current implementation of balancer in gRPC-go, the issues that cannot be solved with the current set of APIs, and proposed solutions for the issues.
+
+## Background
+
+### The current implementation
+
+![before](L9_graphics/bar_before.png)
+
+gRPC talks to balancer directly, but doesn't know about the existence of the resolver. Resolver is essentially a part of balancer from gRPC's point of view. Balancer maintains the pool of resolved addresses and monitors the updates from resolver.
+
+#### Initialization
+
+```go
+grpc.Dial(targetName, grpc.WithBalancer(grpc.Roundrobin(userCustomResolver())))
+```
+
+Upon dialing, gRPC takes the user provided balancer. Users don't provide resolver to gRPC directly. In the example code above, the balancer constructor takes a resolver as an argument. Other balancer implementations may choose to not take the resolver.
+
+gRPC starts the balancer with the target name. The balancer in the example will initialize the resolver as well.
+
+#### Address updates
+
+Balancer gets the address updates from resolver, maintains a pool of resolved addresses, and notifies gRPC of the addresses to which gRPC should create connections. gRPC will create a connection to each of the address, and maintain the connection pool. There will be a gRPC goroutine watching balancer updates and maintaining the connection pool. (Address is the address(IP) + some metadata, so it's possible to create multiple connections to one same server).
+
+Note that the updates in the current resolver API are delta, not the full address list. This is the reason of one of our issues.
+
+##### If balancer is not provided
+
+If balancer is not provided, the target name will be used directly to do net.Dial(). net.Dial() will do the resolving part and return a connection for use.
+
+#### Connection up and down
+
+gRPC notifies the balancer of addresses to which the connection is established or lost. So balancer can make load balancing decisions based on the connection status.
```

Comment:
how will this interact with health checking? if http2 or customized health checker fails, will the connection change state to lost, and will the balancer get notified?

Reply:
gRPC notices a connection is lost when it fails to read-from/write-to the connection. (We have [`keepalive`](https://github.com/grpc/grpc-go/blob/master/keepalive/keepalive.go) package for point-to-point healthcheck).

If there's another custom health checker reporting unhealthy but the connection is actually working, gRPC will not notice that.

Reply:
> If there's another custom health checker reporting unhealthy but the connection is actually working, gRPC will not notice that.

we probably need a hook for that too. Or we have to manually close that connection somehow. 

for quorum based systems, network partitioned server (partitioned from the majority of its peers) is not really usable for some types of RPC although the network between that server and client is OK.

Reply:
In this case, I think it would make more sense for the health checker to communicate to the balancer directly, not through gRPC.

This is what we do in `grpclb`. `grpclb` connects to a remote balancer, and the remote balancer decides which backends to connect and which to disconnect.

Reply:
what you suggest is what we are doing right now. but then down (or done) func will have multiple consumers, and i feel it is less cleaner. i will take a look at grpcLB code. 

thank you.